ethcore/trace/src/types/trace.rs

```diff
@@ -160,7 +160,41 @@ pub struct Call {
 	/// The input data provided to the call.
 	pub input: Bytes,
 	/// The type of the call.
-	pub call_type: Option<CallType>,
+	pub call_type: BackwardsCompatibleOption<CallType>,
+}
+
+/// This is essentially an `Option<T>`,
+/// but with a custom `rlp::Decodable` implementation
+/// which preserves backwards compatibility with
+/// the older encoding (`T`) used in parity-ethereum versions < 2.7.
+/// Note, that this only works if `T` is serialized `as_data`, not `as_list`.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct BackwardsCompatibleOption<T>(pub Option<T>);
+
+impl<T> From<Option<T>> for BackwardsCompatibleOption<T> {
+	fn from(option: Option<T>) -> Self {
+		BackwardsCompatibleOption(option)
+	}
+}
+
+// Encoding is the same as `Option<T>`
+impl<T: Encodable> Encodable for BackwardsCompatibleOption<T> {
+	fn rlp_append(&self, s: &mut RlpStream) {
+		self.0.rlp_append(s);
+	}
+}
+
+// Try to decode it as `T` first, and then as `Option<T>`.
+impl<T: Decodable> Decodable for BackwardsCompatibleOption<T> {
+	fn decode(rlp: &Rlp) -> Result<Self, DecoderError> {
+		if rlp.is_data() {
+			let raw: T = Decodable::decode(rlp)?;
+			Ok(BackwardsCompatibleOption(Some(raw)))
+		} else {
+			let optional: Option<T> = Decodable::decode(rlp)?;
```

Comment:
So if `rlp` is a list here it will fail? Is this a fundamental limitation or can it be fixed later? Not sure I understand why the list case will fail, mind spelling it out?